### DISCUSSION TITLE
Why does riverpod rebuild when the current and new state is the same ?

### DISCUSSION THREAD

**Discussion by Velkamhell97** [rrousselGit/riverpod#2807] [Q&A] [answered]
I made the following example in DartPad and I realize that riverpod only makes the comparison of current and next state when the data is primary as an int or String, but if I have a class as state, and I update to a new value exactly equal to the previous one, in the same way a rebuild is made, even if the class has Equatable for the comparison this continues happening, is this normal ?, in Bloc it is only rebuilds when the new state is different to the previous one. I would like someone to explain me why Riverpod works this way or if I am making a mistake.

https://dartpad.dev/?null_safety=true&id=38ac28be4efeefb363f3647f77c107bd

Thank you very much.!

**Comment by AhmedLSayed9** [comment]
The example you shared doesn't compile and it's using very old Riverpod version.

**Reply by Velkamhell97** [reply]
Sorry, i dont save the last changes, there is the link again.

https://dartpad.dev/?null_safety=true&id=26b87ca48bfc4873be822a9d060137ca

Here is a video with the lastest version of flutter-riverpod

https://github.com/rrousselGit/riverpod/assets/54115437/e49cde85-2ccd-44b2-bf53-162135f5d86d

**Comment by AhmedLSayed9** [comment]
From API docs: By default, the previous and new value are compared using `identical` for performance reasons. If you do not want that, you can override `updateShouldNotify` method to perform a deep comparison of the previous and new values.

Also, You shouldn't use the legacy `StateNotifierProvider` and use the modern `NotifierProvider` that has more features and will support the upcoming features too.

**Reply by Velkamhell97** [reply]
I already try with the `NotifierProvider` class and the behaviour is the same.

https://dartpad.dev/?null_safety=true&id=26b87ca48bfc4873be822a9d060137ca

Can you show me the link where i can find how to override this behaviour and make a deep comparison or give me an example, i have been looking in the documentation but I can't find any information, the `NotifierProvider` class don't have a method that i can override. I see that you can filter rebuilds using `.select` but that is not the way that i want to do. Thank you very much

**Reply by AhmedLSayed9** [reply, marked as answer]
The method you need to override is called `updateShouldNotify`:
```dart
@override
  bool updateShouldNotify(CounterState previous, CounterState next) {
    if (identical(previous, next)) return false; // if they're identical no need to compare with ==
    // Add your == comparision
    return true;
  }
```

**Reply by Velkamhell97** [reply]
Thanks
